### ISSUE TITLE
Android 15 stops syncthing-fork in the background after 6 hours

### ISSUE THREAD

**Issue by mrwm** [Catfriend1/syncthing-android#1147]
### Description of the issue
- Android 15 kills the app in the background after around 6 hours.

Some resources
- https://developer.android.com/about/versions/15/behavior-changes-15#datasync-timeout
- https://developer.android.com/develop/background-work/services/fg-service-timeout

### Reproduction Steps:
1. Set syncthing-fork to Force start ignore run conditions
2. Wait for 6 hours
3. See the error message about syncthing not responding

### Version Information
- App Version: version 1.27.9.0 com.github.catfriend1.syncthingandroid (fdroid)
- Syncthing Version: v1.27.9
- Android Version: Stock Android 15 
- Device manufacturer: Google 
- Device model: Pixel 9 Pro 

### Device platform info
<details><summary><pre>adb shell "getprop | grep ro.product"</pre></summary>
<pre>
[ro.product.ab_ota_partitions]: [abl,bl1,bl2,bl31,boot,dtbo,gcf,gsa,gsa_bl1,init_boot,ldfw,modem,pbl,product,pvmfw,system,system_dlkm,system_ext,tzsw,vbmeta,vbmeta_system,vbmeta_vendor,vendor,vendor_boot,vendor_dlkm,vendor_kernel_boot]
[ro.product.board]: [caiman]
[ro.product.bootimage.brand]: [google]
[ro.product.bootimage.device]: [caiman]
[ro.product.bootimage.manufacturer]: [Google]
[ro.product.bootimage.model]: [Pixel 9 Pro]
[ro.product.bootimage.name]: [caiman]
[ro.product.brand]: [google]
[ro.product.brand_for_attestation]: [google]
[ro.product.build.date]: [Fri Sep 13 17:34:55 UTC 2024]
[ro.product.build.date.utc]: [1726248895]
[ro.product.build.fingerprint]: [google/caiman/caiman:15/AP3A.241005.015/12366759:user/release-keys]
[ro.product.build.id]: [AP3A.241005.015]
[ro.product.build.tags]: [release-keys]
[ro.product.build.type]: [user]
[ro.product.build.version.incremental]: [12366759]
[ro.product.build.version.release]: [15]
[ro.product.build.version.release_or_codename]: [15]
[ro.product.build.version.sdk]: [35]
[ro.product.cpu.abi]: [arm64-v8a]
[ro.product.cpu.abilist]: [arm64-v8a]
[ro.product.cpu.abilist32]: []
[ro.product.cpu.abilist64]: [arm64-v8a]
[ro.product.cpu.pagesize.max]: [16384]
[ro.product.device]: [caiman]
[ro.product.device_for_attestation]: [caiman]
[ro.product.first_api_level]: [34]
[ro.product.locale]: [en-US]
[ro.product.manufacturer]: [Google]
[ro.product.manufacturer_for_attestation]: [Google]
[ro.product.model]: [Pixel 9 Pro]
[ro.product.model_for_attestation]: [Pixel 9 Pro]
[ro.product.name]: [caiman]
[ro.product.name_for_attestation]: [caiman]
[ro.product.odm.brand]: [google]
[ro.product.odm.device]: [caiman]
[ro.product.odm.manufacturer]: [Google]
[ro.product.odm.model]: [Pixel 9 Pro]
[ro.product.odm.name]: [caiman]
[ro.product.product.brand]: [google]
[ro.product.product.device]: [caiman]
[ro.product.product.manufacturer]: [Google]
[ro.product.product.model]: [Pixel 9 Pro]
[ro.product.product.name]: [caiman]
[ro.product.system.brand]: [google]
[ro.product.system.device]: [generic]
[ro.product.system.manufacturer]: [Google]
[ro.product.system.model]: [mainline]
[ro.product.system.name]: [mainline]
[ro.product.system_dlkm.brand]: [google]
[ro.product.system_dlkm.device]: [caiman]
[ro.product.system_dlkm.manufacturer]: [Google]
[ro.product.system_dlkm.model]: [Pixel 9 Pro]
[ro.product.system_dlkm.name]: [caiman]
[ro.product.system_ext.brand]: [google]
[ro.product.system_ext.device]: [caiman]
[ro.product.system_ext.manufacturer]: [Google]
[ro.product.system_ext.model]: [Pixel 9 Pro]
[ro.product.system_ext.name]: [caiman]
[ro.product.vendor.brand]: [google]
[ro.product.vendor.device]: [caiman]
[ro.product.vendor.manufacturer]: [Google]
[ro.product.vendor.model]: [Pixel 9 Pro]
[ro.product.vendor.name]: [caiman]
[ro.product.vendor_dlkm.brand]: [google]
[ro.product.vendor_dlkm.device]: [caiman]
[ro.product.vendor_dlkm.manufacturer]: [Google]
[ro.product.vendor_dlkm.model]: [Pixel 9 Pro]
[ro.product.vendor_dlkm.name]: [caiman]
</pre>
</details>

### Android Log

```sh
Process: com.github.catfriend1.syncthingandroid
java.lang.RuntimeException: Unable to start service com.nutomic.syncthingandroid.service.SyncthingService@6c219ac with null: android.app.ForegroundServiceStartNotAllowedException: Time limit already exhausted for foreground service type dataSync

android.app.ForegroundServiceStartNotAllowedException
at com.nutomic.syncthingandroid.util.Util$$ExternalSyntheticApiModelOutline0.m
at com.nutomic.syncthingandroid.service.NotificationHandler.updatePersistentNotification
at com.nutomic.syncthingandroid.service.SyncthingService.onStartCommand
  at android.app.ActivityThread.handleServiceArgs(ActivityThread.java:5165)
  at android.app.ActivityThread.-$$Nest$mhandleServiceArgs(Unknown Source:0)
  at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2469)
  at android.os.Handler.dispatchMessage(Handler.java:107)
  at android.os.Looper.loopOnce(Looper.java:232)
  at android.os.Looper.loop(Looper.java:317)
  at android.app.ActivityThread.main(ActivityThread.java:8705)
  at java.lang.reflect.Method.invoke(Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:580)
  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:886)
Caused by: android.app.ForegroundServiceStartNotAllowedException: Time limit already exhausted for foreground service type dataSync
  at android.app.ForegroundServiceStartNotAllowedException$1.createFromParcel(ForegroundServiceStartNotAllowedException.java:54)
  at android.app.ForegroundServiceStartNotAllowedException$1.createFromParcel(ForegroundServiceStartNotAllowedException.java:50)
  at android.os.Parcel.readParcelableInternal(Parcel.java:5075)
  at android.os.Parcel.readParcelable(Parcel.java:5057)
  at android.os.Parcel.createExceptionOrNull(Parcel.java:3237)
  at android.os.Parcel.createException(Parcel.java:3226)
  at android.os.Parcel.readException(Parcel.java:3209)
  at android.os.Parcel.readException(Parcel.java:3151)
  at android.app.IActivityManager$Stub$Proxy.setServiceForeground(IActivityManager.java:7193)
  at android.app.Service.startForeground(Service.java:863)
  at com.nutomic.syncthingandroid.util.Util$$ExternalSyntheticApiModelOutline0.m(D8$$SyntheticClass:0)
  at com.nutomic.syncthingandroid.service.NotificationHandler.updatePersistentNotification(NotificationHandler.java:216)
  at com.nutomic.syncthingandroid.service.NotificationHandler.updatePersistentNotification(NotificationHandler.java:109)
  at com.nutomic.syncthingandroid.service.SyncthingService.onStartCommand(SyncthingService.java:306)
  at android.app.ActivityThread.handleServiceArgs(ActivityThread.java:5147)
```

**Comment by daveschroeter** [comment]
Same here. Was hoping this was a beta issue but still present in stable.

**Comment by TomasRiker** [comment]
The main problem is that the app targets Android 15 without being tested on Android 15. Changing `targetSdkVersion` back to 34 should fix this, because the timeout only applies to apps targeting Android 15.

**Comment by chenxiaolong** [comment]
This is where AOSP sets the time limits: https://android.googlesource.com/platform/frameworks/base/+/refs/tags/android-15.0.0_r3/services/core/java/com/android/server/am/ActiveServices.java#3757

If publishing on Google Play doesn't matter, then switching to `FOREGROUND_SERVICE_TYPE_SPECIAL_USE` should do the trick. That type has no time limit. (Android itself has no restrictions on which foreground service types you're allowed to use. Those are Google Play restrictions.)

**Comment by DrSchnagels** [comment]
> This is where AOSP sets the time limits: https://android.googlesource.com/platform/frameworks/base/+/refs/tags/android-15.0.0_r3/services/core/java/com/android/server/am/ActiveServices.java#3757
> 
> If publishing on Google Play doesn't matter, then switching to `FOREGROUND_SERVICE_TYPE_SPECIAL_USE` should do the trick. That type has no time limit. (Android itself has no restrictions on which foreground service types you're allowed to use. Those are Google Play restrictions.)

Oh my god. This gives me so much hope. I was regretting buying another Google product because of this 6 h limit (pixel tablet).
Because with my Pixel phone and tablet.... the 6 h limit is affecting me since the Android 15 update. And it is vvvvvvvvvvvvvvery annoying. Thanks for researching this!
now @Catfriend1 has an option to help us against the 6 h limit, if he has time <3.

**Comment by Catfriend1** [comment]
Thanks for bringing those things to my agenda. Will see if I can take on them. PRs also welcome.

Until we have full Android 15 compat, you can use an older release to get Syncthing-Fork working as you are used to.

**Comment by Catfriend1** [comment]
Please provide feedback if this release fixed the issue:
https://github.com/Catfriend1/syncthing-android/releases/download/v1.28.0.0/com.github.catfriend1.syncthingandroid_github_v1.28.0.0_0af811a5.apk

**Comment by pdelre** [comment]
So far so good. My current uptime is 10h 54m.

**Comment by hachigoro** [comment]
Hello, I have been using 1.28 (from Play Store), but with Android 15 it keeps shutting down after a few hours of use. Thanks for investigating this.

**Comment by Catfriend1** [comment]
See https://github.com/Catfriend1/syncthing-android/issues/1175

**Comment by Catfriend1** [comment]
Maybe solved since v1.29.4+
